<H1>Economic Lot Sizing</H1>

Consider a discrete-time, finite-horizon inventory control problem where you must satisfy a given demand. $d_t$ for each week $t$. This demand can be satisfied by production in any week $t' \le t$. For each week, there is a per unit production cost, a per unit inventory holding cost, and a setup cost. The setup cost is a fixed cost incurred if any production occurs during the week. 

Since the setup cost is paid at most once each week, regardless of whether you produce 1 unit or 1000, you have an incentive to potentially produce enough during a week to cover the demand for multiple weeks. But, producing in excess of the week's demand requires you to hold inventory. This tradeoff between setup and holding costs can be managed through optimization.

<H2>Initial Formulation</H2>
We'll consider two formulations of the problem. The first has fewer decision variables but has a drawback that we'll soon discover.

<H3>Sets and Indices</H3>
* $t \in \{1,\ldots, T\}$: weeks

<H3>Data</H3>
* $c_t$: per unit cost of production in week $t$
* $h_t$: per unit holding cost for inventory remaining in week $t$
* $f_t$: fixed cost of production in week $t$
* $d_t$: demand for week $t$

<H3>Decision Variables</H3>
* $x_t$: binary variable to indicate whether production occurs in week $t$
* $y_t$: production quantity for week $t$
* $s_t$: inventory carried in to period $t$

<H3>Formulation</H3>
\begin{eqnarray}
\min_{x, y, s} && \sum_{t = 1}^T (f_t x_t + c_t y_t + h_t s_t) \nonumber \\
\mbox{s.t.} && y_t \le (\sum_{t' = t}^{T} d_{t'}) x_t, \;\;t = 1,\ldots, T\nonumber \\
&& s_t = s_{t-1} + y_t - d_t,\;\;t = 1,\ldots,T \nonumber \\
&& s_0 = 0 \nonumber \\
&& x_t \in \{0, 1\},\;\;t = 1,\ldots,T \nonumber \\
&& y_t \ge 0,\;\; t = 1,\ldots,T \nonumber \\
&& s_t \ge 0,\;\; t = 1,\ldots,T \nonumber \\
\end{eqnarray}


In [1]:
demands = [20, 50, 10, 50, 50, 10, 20, 40, 20, 30]

In [2]:
production_costs = [10, 10, 10, 10, 10, 10, 10, 10, 10, 10]

In [3]:
fixed_costs = [100, 100, 100, 100, 100, 100, 100, 100, 100, 100]

In [4]:
holding_costs = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

In [5]:
import gurobipy as grb
GRB = grb.GRB

def get_lot_sizes(demands, production_costs, fixed_costs, holding_costs):
    m = grb.Model()
    
    weeks = range(len(demands))
    production_quantity = [m.addVars(name='quantity.' + str(week))
                           for week in weeks]
    holding_quantity = [m.addVar(name='hold.' + str(week))
                        for week in weeks]
    to_produce = [m.addVar(name='to_produce.' + str(week),
                           vtype=GRB.BINARY)
                  for week in weeks]
    m.update()
    
    for week in weeks:
        #Add a constraint that ensures that production_quantity is only nonzero in weeks where to_produce is also nonzero
        m.addConstr(...)
        carried_inventory = holding_quantity[week-1] if week > 0 else 0
        #Add an inventory flow constraint that maintains an appropriate holding_quantity value for each week
        m.addConstr(...)
    
    total_production_cost = grb.quicksum(production_costs[week]*production_quantity[week]
                                         for week in weeks)
    total_fixed_cost = grb.quicksum(fixed_costs[week]*to_produce[week]
                                    for week in weeks)
    total_holding_cost = grb.quicksum(holding_costs[week]*holding_quantity[week]
                                      for week in weeks)
    m.setObjective(total_production_cost + total_fixed_cost + total_holding_cost)

    m.optimize()
    
    if m.Status == GRB.OPTIMAL:
        for week, var in enumerate(production_quantity):
            if var.X > 1e-4:
                print ("produce", var.X, "units in week", week)

    

In [6]:
get_lot_sizes(demands, production_costs, fixed_costs, holding_costs)

Using license file C:\Users\liaml\gurobi.lic
Academic license - for non-commercial use only
Gurobi Optimizer version 9.0.2 build v9.0.2rc0 (win64)
Optimize a model with 20 rows, 30 columns and 49 nonzeros
Model fingerprint: 0x00e4c5e5
Variable types: 20 continuous, 10 integer (10 binary)
Coefficient statistics:
  Matrix range     [1e+00, 3e+02]
  Objective range  [1e+00, 1e+02]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+01, 5e+01]
Presolve removed 3 rows and 4 columns
Presolve time: 0.00s
Presolved: 17 rows, 26 columns, 42 nonzeros
Variable types: 17 continuous, 9 integer (9 binary)

Root relaxation: objective 3.375304e+03, 21 iterations, 0.00 seconds

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0 3375.30360    0    7          - 3375.30360      -     -    0s
H    0     0                    3930.0000000 3375.30360  14.1%     -    0s
     0     0 3520.13987 

It is therefore optimal to produce 80 units in week 0 to cover demand through week 2, to produce 130 units in week 3 to cover demand for weeks 3 through 6, and to produce 90 units in week 7 to cover demand for weeks 7 through 9.

Note that the LP relaxation produced a lower bound on the cost of 3375.304, but ultimately that lower bound could not be acheived as the true minimum cost was 3580. It is possible to formulate the problem in a way that the LP relaxation value is tight, and we can obtain the optimal integer solution immediately without resorting to branch and bound.

<H2>A Tighter Formulation</H2>

We'll consider a reformulation of the problem which has quadratically many variables, but that produces an LP relaxation that is tight. While the problem size will increase, this model will scale better since it produces an integer feasible solution immediately upon solving the LP relaxation.

The trick here is to note that production in week $t$ may be used to satisfy demand in any week $t' \ge t$. We can define the production variables to include both a production week $t$ and a demand week $t'$. We'll need to adjust the cost of these variables to include both the variable cost of production and the cost to hold inventory from week $t$ into week $t'$. This adjusted cost is simply $c_t + \sum_{i = t}^{t'-1} h_i$.

<H3>Data</H3>
* $\bar{c}_{tt'}$: cost to produce a unit in week $t$ and hold it as inventory until week $t'$
* $f_t$: fixed cost of production in week $t$

<H3>Decision Variables</H3>
* $x_t$: binary variable to indicate whether production occurs in week $t$
* $y_{tt'}$: number of units produced in week $t$ to satisfy demand in week $t' \ge t$

<H3>Formulation</H3>
\begin{eqnarray}
\min_{x, y} && \sum_{t = 1}^T \sum_{t'=t}^T \bar{c}_{tt'} y_{tt'} + \sum_{t = 1}^T f_t x_t \nonumber \\
\mbox{s.t.} && \sum_{t = 1}^{t'} y_{tt'} = d_{t'},\;\;t' = 1,\ldots, T \nonumber \\
&& y_{tt'} \le d_{t'}x_{t},\;\;t' = 1,\ldots, T,\;\;t=1,\ldots,t' \nonumber \\
&& x_t \in \{0, 1\},\;\;t = 1,\ldots, T \nonumber \\
&& y_{tt'} \ge 0,\;\;t = 1,\ldots, T,\;\;t' = t,\ldots,|T|. \nonumber
\end{eqnarray}

In [7]:
def get_lot_sizes(demands, production_costs, fixed_costs, holding_costs):
    m = grb.Model()
    weeks = range(len(demands))
    
    production_quantity = {}
    for production_week in weeks:
        stuff=m.addVars([production_week], weeks[production_week:])
        stuff[0][0].obj = 
        for demand_week in weeks[production_week:]:
            # compute the cost to produce in production_week and hold until demand_week
            cost = ...
            # add a decision variable with that cost
            var = ...
            production_quantity[production_week, demand_week] = var

    # add a binary variable for each week to indicate whether production occurs in that week
    to_produce = [...]

    m.update()

    for demand_week in weeks:
        # add a constraint to ensure that demand is satisfied in demand_week
        # the production can come from any production_week = 1,...,demand_week
        m.addConstr(...)
    
    for (production_week, demand_week), production_var in production_quantity.items():
        # add a constraint to enforce that production_var is 0 unless the corresponding to_produce variable is 1
        m.addConstr(...)
    
    m.optimize()
    
    for demand_week, var in enumerate(to_produce):
        print (demand_week, var.X)
        
    m.write('lot_sizing.lp')
    m.write('lot_sizing.sol')